ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/HiveCalciteUtil.java

```diff
@@ -683,6 +687,20 @@ public static ImmutableList<RexNode> getPredsNotPushedAlready(Set<String> predic
     predicatesToExclude.addAll(predicatesInSubtree);
     return newConjuncts.build();
   }
+  
+  private static RexNode simplify(RexSimplify simplifier, RexNode node) {
+    RexNode result = node;
+    int maxTries = 5;
+    for (int i = 0; i < maxTries; i++) {
+      RexNode simplified = simplifier.simplify(result);
+      if (simplified.equals(result)) {
+        break;
+      }
+      result = simplified;
+    }
+    
+    return result;
```

Comment:
This part indicates that there is some part of flakiness/nondeterminism in the mix. Moreover, the simplifier is rather costly so calling it multiple times can have an impact on performance. Why do we need to do this? What is the expression that requires multiple rounds of simplification? This probably deserves logging a dedicated CALCITE ticket. Unfortunately this cannot be committed as such we need to work out a solution.

Reply:
Yes, this is rather crude, and I will remove it. I just wanted to test that if the RexNode is completely simplified at this point, then we won't run into the infinite for loop issue - this is explained by Krisztian [here](https://github.com/apache/hive/pull/5196#discussion_r1660939490).

There is already a ticket in Calcite to fix this: [CALCITE-6453](https://issues.apache.org/jira/browse/CALCITE-6453).

So the issue is really with certain CASTs not getting reduced, and we have to call simplify twice to reduce it.